### DISCUSSION TITLE
Unable to infer GraphQL type from TypeScript reflection system

### DISCUSSION THREAD

**Discussion by hmz22** [MichalLytek/type-graphql#1320] [Q&A]
**Describe the Bug**
I create this model

```
import { getModelForClass } from "@typegoose/typegoose";
import { GraphQLObjectID } from "graphql-scalars";
import { Field, ObjectType } from "type-graphql";
import { publicConnection } from "../../startup/db";

@ObjectType()
export class Member {
  @Field(() => GraphQLObjectID)
  _id!: string;

  @Field()
  firstname?: string;

  @Field()
  lastname?: string;
}

export const MemberModel = getModelForClass(Member, {
  existingConnection: publicConnection,
  schemaOptions: {
    timestamps: true,
  },
});
```
and this my resolver:

```
import { Query, Resolver } from "type-graphql";
import { Member, MemberModel } from "./member.model";

@Resolver(() => Member)
export default class MemberResolver {
  @Query(() => [Member])
  async members() {
    return await MemberModel.find({});
  }
}

```

after run project show this error in console

```
D:\Projects\Tests\node-ts-sample-decorator2\node_modules\type-graphql\dist\helpers\findType.js:19
        throw new errors_1.NoExplicitTypeError(prototype.constructor.name, propertyKey, parameterIndex, argName);
              ^
Error: Unable to infer GraphQL type from TypeScript reflection system. You need to provide explicit type for 'firstname' of 'Member' class.
```

when I change `@Field() to @Field(type => String)` this error gone but in your document no need add this

- OS: Windows 11
- Node 14.18.3
- Package version [1.2.0-rc.1]
- TypeScript version 4.7.4

**Comment by hmz22** [comment]
> **Describe the Bug** I create this model
> 
> ```
> import { getModelForClass } from "@typegoose/typegoose";
> import { GraphQLObjectID } from "graphql-scalars";
> import { Field, ObjectType } from "type-graphql";
> import { publicConnection } from "../../startup/db";
> 
> @ObjectType()
> export class Member {
>   @Field(() => GraphQLObjectID)
>   _id!: string;
> 
>   @Field()
>   firstname?: string;
> 
>   @Field()
>   lastname?: string;
> }
> 
> export const MemberModel = getModelForClass(Member, {
>   existingConnection: publicConnection,
>   schemaOptions: {
>     timestamps: true,
>   },
> });
> ```
> 
> and this my resolver:
> 
> ```
> import { Query, Resolver } from "type-graphql";
> import { Member, MemberModel } from "./member.model";
> 
> @Resolver(() => Member)
> export default class MemberResolver {
>   @Query(() => [Member])
>   async members() {
>     return await MemberModel.find({});
>   }
> }
> ```
> 
> after run project show this error in console
> 
> ```
> D:\Projects\Tests\node-ts-sample-decorator2\node_modules\type-graphql\dist\helpers\findType.js:19
>         throw new errors_1.NoExplicitTypeError(prototype.constructor.name, propertyKey, parameterIndex, argName);
>               ^
> Error: Unable to infer GraphQL type from TypeScript reflection system. You need to provide explicit type for 'firstname' of 'Member' class.
> ```
> 
> when I change `@Field() to @Field(type => String)` this error gone but in your document no need add this
> 
> * OS: Windows 11
> * Node 14.18.3
> * Package version [1.2.0-rc.1]
> * TypeScript version 4.7.4

I forget say I use esbuild for run this program

**Reply by MichalLytek** [reply]
That's the root cause. See #55

**Reply by samrith-s** [reply]
@MichalLytek I couldn't gather much from the really long discussion in that issue. Any leads on how to solve it for `esbuild`?

**Comment by hmz22** [comment]
I install @anatine/esbuild-decorators and solved my problem because esbuild not support decorator.

**Reply by samrith-s** [reply]
I tried it with the plugin, still same issue. This is my `esbuild.config.js`:
```js
/* eslint-disable @typescript-eslint/no-var-requires */
const { esbuildDecorators } = require('@anatine/esbuild-decorators');

module.exports = () => ({
  packager: 'yarn',
  packagerOptions: {
    scripts: [
      'rm -rf node_modules',
      'cp ../../prisma/schema.prisma src/schema.prisma',
    ],
  },
  keepOutputDirectory: true,
  external: [
    'cache-manager',
    '@nestjs/websockets',
    '@nestjs/microservices',
    'superagent-proxy',
    'apollo-server-fastify',
    'class-transformer',
    '@apollo/federation',
    'fsevents',
  ],
  plugins: [esbuildDecorators()],
});
```

**Reply by samrith-s** [reply]
My bad! I was running the wrong command. I needed to invoke the local function, not the errored one which has been deployed:
```diff
- sls invoke -f main
+ sis invoke local -f main
```

**Comment by 0xborisgregor** [comment]
For anyone who already configured 
```json
{
  "emitDecoratorMetadata": true,
  "experimentalDecorators": true
}
```
 in `tsconfig.json` but still run into `NoExplicitTypeError`, you might be using `tsx` which doesn't support `emitDecoratorMetadata`: [tsx compiler limitations](https://tsx.is/typescript#compiler-limitations)
